certdehydrate/certdehydrate.go

```diff
@@ -0,0 +1,231 @@
+package certdehydrate
+
+import "bytes"
+import "crypto/sha256"
+import "crypto/x509/pkix"
+import "encoding/base64"
+import "encoding/binary"
+import "fmt"
+import "math/big"
+import "time"
+
+import "github.com/hlandau/ncdns/x509"
+
+// TODO: add a version field
+type DehydratedCertificate struct {
+	PubkeyB64 string
+	NotBeforeScaledInt int64
+	NotAfterScaledInt int64
+	SignatureAlgorithmInt int64
+	SignatureB64 string
+}
+
+func (dehydrated DehydratedCertificate) SerialNumber(name string) ([]byte, error){
+	
+	nameHash := sha256.Sum256([]byte(name))
+	
+	pubkeyBytes, err := base64.StdEncoding.DecodeString(dehydrated.PubkeyB64)
+	if err != nil {
+		return nil, fmt.Errorf("Dehydrated cert pubkey is not valid base64: %s", err)
+	}
+	pubkeyHash := sha256.Sum256(pubkeyBytes)
+	
+	notBeforeScaledBuf := new(bytes.Buffer)
+	err = binary.Write(notBeforeScaledBuf, binary.BigEndian, dehydrated.NotBeforeScaledInt)
+	if err != nil {
+		fmt.Println("binary.Write failed:", err)
+	}
+	notBeforeHash := sha256.Sum256(notBeforeScaledBuf.Bytes())
+	
+	notAfterScaledBuf := new(bytes.Buffer)
+	err = binary.Write(notAfterScaledBuf, binary.BigEndian, dehydrated.NotAfterScaledInt)
+	if err != nil {
+		fmt.Println("binary.Write failed:", err)
+	}
+	notAfterHash := sha256.Sum256(notAfterScaledBuf.Bytes())
+	
+	//notBeforeHash := sha256.Sum256([]byte(dehydrated.NotBeforeScaledInt))
+	//notAfterHash := sha256.Sum256([]byte(dehydrated.NotAfterScaledInt))
+		
+	serialHash := sha256.New()
+	serialHash.Write(nameHash[:])
+	serialHash.Write(pubkeyHash[:])
+	serialHash.Write(notBeforeHash[:])
+	serialHash.Write(notAfterHash[:])
+	
+	// 19 bytes will be less than 2^159, see https://crypto.stackexchange.com/a/260
+	return serialHash.Sum(nil)[0:19], nil
+}
+
+func (dehydrated DehydratedCertificate) String() string {
```

Comment:
You could use encoding/json to do this: serialize a []interface{} containing the values. This works, though it could have a vulnerability if bad data gets into the base64 fields somehow. At any rate, there's some whitespace in here you can remove to minimise the size.